### PR TITLE
Package exec implements execution of a graph of Action and Events.

### PR DESCRIPTION
**Pull request by bowei**
- Actions produce Events when they complete.
- Events fulfill dependencies of Actions so they can be executed.

```
go test -cover ./pkg/cloud/resgraph/exec 
ok  	github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/resgraph/exec	0.028s	coverage: 93.0% of statements
```

### CODE REVIEW

File: pkg/cloud/rgraph/exec/event.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+)
+
+// Event occurs when an Action completes. Events can signal dependent Actions to
+// be available for execution.
+type Event interface {
+	// Equal returns true if this event is == to other.
+	Equal(other Event) bool
+	// String implements Stringer.
```
Comment (briantkennedy) [inline pkg/cloud/rgraph/exec/event.go:30]: use `fmt.Stringer`?

File: pkg/cloud/rgraph/exec/executor.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+package exec
+
+import (
+	"context"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+)
+
+type ExecutorResult struct {
```
Comment (briantkennedy) [inline pkg/cloud/rgraph/exec/executor.go:25]: mild go nit: this is package `exec`, so the `Executor` prefix may be redundant on some of the structs.  Any thoughts on `exec.UtorResult`? (I'm kidding, but `exec.ExecutorResult` does have the echo going on)

**Review by briantkennedy — APPROVED** [review]

File: pkg/cloud/resgraph/exec/executor.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		DryRun:        false,
+		ErrorStrategy: StopOnError,
+	}
+}
+
+// ExecutorConfig for the executor implementation.
+type ExecutorConfig struct {
+	Tracer        Tracer
+	DryRun        bool
+	ErrorStrategy ErrorStrategy
```
[the remark this answers is not in the document]
Comment (bowei) [reply pkg/cloud/resgraph/exec/executor.go]: Any suggestions? The string-based constants don't really work with this.

File: pkg/cloud/rgraph/exec/executor.go (hunk 2 above)
Comment (bowei) [reply pkg/cloud/rgraph/exec/executor.go:25]: fixed

File: pkg/cloud/resgraph/exec/executor_serial.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	if runErr == nil {
+		ex.result.Completed = append(ex.result.Completed, a)
+	} else {
+		ex.result.Errors = append(ex.result.Errors, ActionWithErr{Action: a, Err: runErr})
+		switch ex.config.ErrorStrategy {
+		case ContinueOnError:
+		case StopOnError:
+			return fmt.Errorf("serialExecutor: stopping execution (got %v)", runErr)
+		default:
+			return fmt.Errorf("serialExecutor: invalid ErrorStrategy %q", ex.config.ErrorStrategy)
```
[the remark this answers is not in the document]
Comment (bowei) [reply pkg/cloud/resgraph/exec/executor_serial.go]: Done

**Pull request merged**